### ISSUE TITLE
Bad ReadGroup formatting in records

### ISSUE THREAD

**Issue by Poshi** [CGATOxford/UMI-tools#285]
For some reason, while using `umi_tools group`, the read group tags get wrongly formatted on one of the ends (the first in pair). As stated in [SAMtags](https://samtools.github.io/hts-specs/SAMtags.pdf), the standard `RG` tag mus be of type `Z`. But in the output of the command, that type is changed to `A`. That leads to error when validating the file and doing other analyses with some tools, like Picard.

Here is a sample input and output showing the wrong behavior:
### Input
    @HD	VN:1.4	SO:coordinate
    @SQ	SN:chr1	LN:248956422
    @RG	ID:1	CN:HOME	DT:2018-10-03	FO:N	LB:Library	PG:GEM	PL:ILLUMINA	PU:HWCMGBBXX_2_19UDI-idt-UMI	SM:Sample
    @PG	ID:GEM	PN:gem-mapper	CL:/apps/GEM/3.6/AVX2/GCC/bin/gem-mapper --threads 8 --sam-read-group-header @RG\tID:1\tCN:HOME\tDT:2018-10-03\tFO:N\tLB:Library\tPG:GEM\tPL:ILLUMINA\tPU:HWCMGBBXX_2_19UDI-idt-UMI\tSM:Sample --clipping=masked -I /project/production/Organisms/Homo_sapiens/GRCh38.hl/gem3/hsapiens.GRCh38.hl.gem -1 umi_test.bam_1.trimmed.fastq -2 umi_test.bam_2.trimmed.fastq	VN:v3.6-dirty-release
    J00148:119:HWCMGBBXX:2:2103:1824:13517	99	chr1	942079	60	101M	=	942227	249	GGCCCTTTACGGGAACGGGGGCGGGGGGGACGCCGCTCATTGCGCTGCCGTCCACAGGGAGCTGCGTCAGCCGCCCCCCTTGCTGTCGTCTCAGAATGCCC	AA-FFFFJJJJJJJFFJJJJFJFJFJJJJJJJJJJJJJJJJJJ7FJ<JJJFJJJJJJJJJJJJJF7-AFJFJJJ<JAFJJF7JFF<JF-7-77AFFJFF-A	RG:Z:1	NM:i:4	MD:Z:2G62C22C1G10	AS:i:81	RX:Z:GACAACATG	QX:Z:JAJFJJFJJ
    J00148:119:HWCMGBBXX:2:2103:1824:13517	147	chr1	942227	60	101M	=	942079	-249	CCCTCGGCTCTGTGCCAGACCCCAGGTGAGGAGGCGGGTGCGCATCCCCTGGGAGCCCGCGTGGAGGCTCGCGGACCCGGCCCTGCCCCTGTCGGAGCCGA	JJFFFAFAFFJJFF-JAFJJFJ7JJJJJJJJJJJJFFJJJJJAF<JFAAFJJAJJJJJJJJJJJJJJJJJJJJJJJJFJFFA<F<JFJ<JJFJJJJFFAA<	RG:Z:1	NM:i:0	MD:Z:101	AS:i:101	RX:Z:GACAACATG	QX:Z:JAJFJJFJJ
### Command
    umi_tools group -I test.sam -S test.group.sam --paired --extract-umi-method tag --umi-tag RX --output-bam --output-unmapped --no-sort-output --log2stderr
### Output
    @HD	VN:1.4	SO:coordinate
    @SQ	SN:chr1	LN:248956422
    @RG	ID:1	CN:HOME	DT:2018-10-03	FO:N	LB:Library	PG:GEM	PL:ILLUMINA	PU:HWCMGBBXX_2_19UDI-idt-UMI	SM:Sample
    @PG	ID:GEM	PN:gem-mapper	CL:/apps/GEM/3.6/AVX2/GCC/bin/gem-mapper --threads 8 --sam-read-group-header @RG\tID:1\tCN:HOME\tDT:2018-10-03\tFO:N\tLB:Library\tPG:GEM\tPL:ILLUMINA\tPU:HWCMGBBXX_2_19UDI-idt-UMI\tSM:Sample --clipping=masked -I /project/production/Organisms/Homo_sapiens/GRCh38.hl/gem3/hsapiens.GRCh38.hl.gem -1 umi_test.bam_1.trimmed.fastq -2 umi_test.bam_2.trimmed.fastq	VN:v3.6-dirty-release
    J00148:119:HWCMGBBXX:2:2103:1824:13517	147	chr1	942227	60	101M	=	942079	-249	CCCTCGGCTCTGTGCCAGACCCCAGGTGAGGAGGCGGGTGCGCATCCCCTGGGAGCCCGCGTGGAGGCTCGCGGACCCGGCCCTGCCCCTGTCGGAGCCGA	JJFFFAFAFFJJFF-JAFJJFJ7JJJJJJJJJJJJFFJJJJJAF<JFAAFJJAJJJJJJJJJJJJJJJJJJJJJJJJFJFFA<F<JFJ<JJFJJJJFFAA<	RG:Z:1	NM:i:0	MD:Z:101	AS:i:101	RX:Z:GACAACATG	QX:Z:JAJFJJFJJ
    J00148:119:HWCMGBBXX:2:2103:1824:13517	99	chr1	942079	60	101M	=	942227	249	GGCCCTTTACGGGAACGGGGGCGGGGGGGACGCCGCTCATTGCGCTGCCGTCCACAGGGAGCTGCGTCAGCCGCCCCCCTTGCTGTCGTCTCAGAATGCCC	AA-FFFFJJJJJJJFFJJJJFJFJFJJJJJJJJJJJJJJJJJJ7FJ<JJJFJJJJJJJJJJJJJF7-AFJFJJJ<JAFJJF7JFF<JF-7-77AFFJFF-A	RG:A:1	NM:i:4	MD:Z:2G62C22C1G10	AS:i:81	RX:Z:GACAACATG	QX:Z:JAJFJJFJJ	UG:i:0	BX:Z:GACAACATG
You can see how the two tags RG:Z:1 from the input mutates into a RG:Z:1 and a RG:A:1 in the output.

**Comment by IanSudbery** [comment]
This is a `pysam` bug, and has been fixed in the latest version of pysam. There is a work around we can do in umi-tools to prevent this, but in the meantime, the easiest way to fix this will be to update your pysam.

@TomSmithCGAT 

This is caused by the use of `read.tags += .....`

I think the tags are being recreated and automatic type guessing is happening. It can be fixed by subsituting the above for `read.set_tag`. I'll do this and submit as a PR.

**Comment by Poshi** [comment]
You are right! I updated from 0.10.0 to 0.15.1 and it looks like it is working properly now. When I installed umi_tools I didn't realized that the package was not upgraded.

Despite the push request you are proposing, another one could be to update the requirements.txt to ask for the minimum version where this issue is solved. Right now it is

    pysam>=0.8.4

Not sure which should be the minimum version where it is solved.
